uint8 start_i
uint8 start_j
uint8 end_i
uint8 end_j
